client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -43,17 +46,44 @@ impl AddrCache {
 			return;
 		}
 
+		addresses.sort_unstable_by(|a, b| a.as_ref().cmp(b.as_ref()));
+
 		// Insert into `self.peer_id_to_authority_id`.
 		let peer_ids = addresses.iter()
 			.map(|a| peer_id_from_multiaddr(a))
 			.filter_map(|peer_id| peer_id);
-		for peer_id in  peer_ids {
-			self.peer_id_to_authority_id.insert(peer_id, authority_id.clone());
+		for peer_id in peer_ids.clone() {
+			let former_auth = match self.peer_id_to_authority_id.insert(peer_id, authority_id.clone()) {
+				Some(a) if a != authority_id => a,
+				_ => continue,
+			};
+
+			// PeerId was associated to a different authority id before.
+			// Remove corresponding authority from `self.authority_id_to_addresses`.
+			let former_auth_addrs = match self.authority_id_to_addresses.get_mut(&former_auth) {
+				Some(a) => a,
+				None => { debug_assert!(false); continue }
```

Comment:
I've heard we have a new policy to include `qed` proofs into debug_asserts as well, but I won't block the PR on that.
Something like
> "the address mapping is bidirectional, if insert returned `Some` on one map, it must return `Some` on the other"

Reply:
Wait, really? This doesn't make sense. I use `debug_assert` *because* they can't panic in production.

Reply:
I might be confusing this with parachain's policy, because we have debug assertions enabled for rococo or even debug_asserts in runtime, so nevermind.